cpp/src/arrow/array/validate.cc

```diff
@@ -64,6 +64,13 @@ struct ValidateArrayVisitor {
     return Status::OK();
   }
 
+  Status Visit(const Decimal256Array& array) {
```

Comment:
Hmm, we could have a `BaseDecimalArray` class like we already have `BaseListArray` and `BaseBinaryArray`.

Reply:
Yes, I tried to make this work, and at the moment making this seems like it would make this change bigger then I would feel comfortable with.  There are a lot of type_traits that have confusing hierarchies (is_primitive and is_binary_like both would include Decimal and SFINAE doesn't work out well, so it would be an intrusive change).

Reply:
FWIW, https://github.com/apache/arrow/pull/8417/files is probably what some of it would look like but I haven't reviewed it fully.

Reply:
Ok.